### PR TITLE
PERF: maybe_promote

### PR DESCRIPTION
**Pull request by jbrockmendel**
Test runtime for pandas/tests/indexing/multiindex/test_indexing_slow.py decreases from 30s to 14.5s.

See discussion around https://github.com/pandas-dev/pandas/pull/39692#issuecomment-1441051740

### CODE REVIEW

File: pandas/core/dtypes/cast.py [hunk 1]
```diff
@@ -556,6 +557,13 @@ def ensure_dtype_can_hold_na(dtype: DtypeObj) -> DtypeObj:
     return dtype
 
 
+_canonical_nans = {
+    np.datetime64: np.datetime64("NaT", "ns"),
```
Comment (mroeschke) [inline pandas/core/dtypes/cast.py:561]: Will it always be correct to assume `"ns"`?

Comment (jbrockmendel) [reply pandas/core/dtypes/cast.py:561]: updated so as to retain identical fill_value in object-result cases

**Comment by jbrockmendel** [comment]
our asv site appears to be done, but locally a lot of the regressions im seeing since 1.5.0 are for `groupby.GroupByMethods.time_dtype_as_field('datetime', ...)` and profiling points at maybe_promote so im guessing this will address a bunch of those

cc @phofl

File: pandas/core/dtypes/cast.py [hunk 2]
```diff
@@ -577,18 +585,29 @@ def maybe_promote(dtype: np.dtype, fill_value=np.nan):
...
-        return _maybe_promote_cached(
+        dtype, fill_value = _maybe_promote_cached(
             dtype, fill_value, type(fill_value)  # type: ignore[arg-type]
         )
     except TypeError:
         # if fill_value is not hashable (required for caching)
-        return _maybe_promote(dtype, fill_value)
+        dtype, fill_value = _maybe_promote(dtype, fill_value)
+
+    if dtype == _dtype_obj and orig is not None:
```
Comment (phofl) [inline pandas/core/dtypes/cast.py:607]: I am not totally sure that I understand why you are checking for None here? The default is np.nan, shouldn't we check for np.nan instead?

Comment (jbrockmendel) [reply pandas/core/dtypes/cast.py:607]: converting None to nan is one that we're supposed to do.  other nan-likes should be retained

Comment (phofl) [reply pandas/core/dtypes/cast.py:607]: got you, thx.

**Review by phofl — APPROVED** [review]

**Comment by phofl** [comment]
thx @jbrockmendel

**Pull request merged**